polygeist/tools/cgeist/Lib/CGStmt.cc

```diff
@@ -1135,6 +1135,27 @@ ValueCategory MLIRScanner::VisitReturnStmt(clang::ReturnStmt *stmt) {
       else if (val.getType().isa<LLVM::LLVMPointerType>() &&
                postTy.isa<MemRefType>())
         val = builder.create<polygeist::Pointer2MemrefOp>(loc, postTy, val);
+      if (auto postMemRefType = postTy.dyn_cast<MemRefType>()) {
+        auto valMemRefType = val.getType().dyn_cast<MemRefType>();
+        assert(valMemRefType && "Expecting val to be memref");
+        assert(postMemRefType.getElementType() ==
+                   valMemRefType.getElementType() &&
+               "Return types mismatch");
+        if (postMemRefType.getMemorySpaceAsInt() !=
+            valMemRefType.getMemorySpaceAsInt()) {
+          val = builder.create<polygeist::Memref2PointerOp>(
+              loc,
+              LLVM::LLVMPointerType::get(valMemRefType.getElementType(),
+                                         valMemRefType.getMemorySpaceAsInt()),
+              val);
+          val = builder.create<LLVM::AddrSpaceCastOp>(
+              loc,
+              LLVM::LLVMPointerType::get(postMemRefType.getElementType(),
+                                         postMemRefType.getMemorySpaceAsInt()),
+              val);
+          val = builder.create<polygeist::Pointer2MemrefOp>(loc, postTy, val);
+        }
```

Comment:
The type of the returned value and the return type of the function should match at this point, right? If they do not fully match (i.e. same element type and address space) something else went wrong ... 

Conceptually a `return` stmt is equivalent to a store operation (the return implicitly stores the value to be returned into a tmp variable and returns that tmp to the caller). In `ValueCategory::store` this PR correctly asserts that both the elem. type and address space match, same assumption should be true here or we are getting into inconsistencies.

Reply:
`_ZNK4sycl3_V18accessorIiLi1ELNS0_6access4modeE1025ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi1EvEERiNS0_2idILi1EEE` is the function that require this change.

Its `ReturnStmt` contains a `AddressSpaceConversion`, added code to check it is indeed that case before creating a `AddrSpaceCastOp`.

From clang:
```
define linkonce_odr dso_local spir_func noundef align 4 dereferenceable(4) i32 addrspace(4)* @_ZNK4sycl3_V18accessorIiLi1ELNS0_6access4modeE1025ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi1EvEERiNS0_2idILi1EEE(%"class.sycl::_V1::accessor" addrspace(4)* noundef align 8 dereferenceable_or_null(32) %this, %"class.sycl::_V1::id"* noundef byval(%"class.sycl::_V1::id") align 8 %Index) #3 comdat align 2 {
...
  %call2 = call spir_func noundef i32 addrspace(1)* @_ZNK4sycl3_V18accessorIiLi1ELNS0_6access4modeE1025ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEE15getQualifiedPtrEv(%"class.sycl::_V1::accessor" addrspace(4)* noundef align 8 dereferenceable_or_null(32) %this1) #10
  %arrayidx = getelementptr inbounds i32, i32 addrspace(1)* %call2, i64 %3
  %arrayidx.ascast = addrspacecast i32 addrspace(1)* %arrayidx to i32 addrspace(4)*
  ret i32 addrspace(4)* %arrayidx.ascast
}
```
`performAddrSpaceCast` called from `ReturnStmt` inserted the `addrspacecast`.
```
#11 0x0000000004be5ef3 clang::CodeGen::TargetCodeGenInfo::performAddrSpaceCast(clang::CodeGen::CodeGenFunction&, llvm::Value*, clang::LangAS, clang::LangAS, llvm::Type*, bool) const /iusers/waihungt/llvm/clang/lib/CodeGen/TargetInfo.cpp:536:1
#12 0x0000000004e6fabc clang::CodeGen::CodeGenFunction::EmitCastLValue(clang::CastExpr const*) /iusers/waihungt/llvm/clang/lib/CodeGen/CGExpr.cpp:4966:60
#13 0x0000000004e5acc8 clang::CodeGen::CodeGenFunction::EmitLValue(clang::Expr const*) /iusers/waihungt/llvm/clang/lib/CodeGen/CGExpr.cpp:1453:44
#14 0x0000000004e564e3 clang::CodeGen::CodeGenFunction::EmitReferenceBindingToExpr(clang::Expr const*) /iusers/waihungt/llvm/clang/lib/CodeGen/CGExpr.cpp:619:3
#15 0x0000000004946489 clang::CodeGen::CodeGenFunction::EmitReturnStmt(clang::ReturnStmt const&) /iusers/waihungt/llvm/clang/lib/CodeGen/CGStmt.cpp:1341:24
```